lib/dotcom/system_status/subway.ex

```diff
@@ -0,0 +1,371 @@
+  # field. Exactly how it does this depends on whether
+  # grouped_statuses has one entry or more.
+  defp nest_grouped_statuses_under_branches(grouped_statuses)
+
+  # If grouped_statuses has one entry, then that means that all of the
+  # branches have the same status, which means we don't need to
+  # specify any branches.
+  defp nest_grouped_statuses_under_branches([{statuses, _}]) do
+    [branch_with_statuses_entry(statuses)]
+  end
+
+  # If grouped_statuses has more than one entry, then we do need to
+  # specify the branches for each collection of statuses.
+  defp nest_grouped_statuses_under_branches(grouped_statuses) do
+    grouped_statuses
+    |> Enum.map(fn {statuses, entries} ->
+      branch_ids = entries |> Enum.map(& &1.route_id) |> Enum.sort()
+
+      branch_with_statuses_entry(statuses, branch_ids)
+    end)
+  end
+
+  # Sorts green line branches first by alert status (that is, "Normal
+  # Service" should come after any other alerts), and then by branch
+  # ID (so that, say statuses for "Green-B" should come ahead of
+  # "Green-C").
+  defp sort_branches(branches_with_statuses) do
+    branches_with_statuses
+    |> Enum.sort_by(fn %{status_entries: status_entries, branch_ids: branch_ids} ->
+      {status_sort_order(status_entries), branch_ids}
+    end)
+  end
+
+  defp status_sort_order([%{time: :current, status: :normal}]), do: 1
+  defp status_sort_order(_), do: 0
+
+  defp branches_with_statuses(route_id, alerts, time) do
+    route_id
+    |> statuses_for_route(alerts, time)
+    |> branch_with_statuses_entry()
+    |> then(&[&1])
+  end
+
+  defp mattapan_branches_with_statuses(alerts, time) do
+    "Mattapan"
+    |> alerts_for_route(alerts)
+    |> case do
+      [] ->
+        []
+
+      mattapan_alerts ->
+        mattapan_statuses = mattapan_alerts |> alerts_to_statuses(time)
+
+        [branch_with_statuses_entry(mattapan_statuses, ["Mattapan"])]
+    end
+  end
+
+  # Exchanges a route_id (a line_id or a branch_id - anything that
+  # might correspond to an alert) for a map with that route_id and the
+  # statuses affecting that route.
+  defp add_statuses_for_route(route_id, alerts, time) do
+    %{
+      route_id: route_id,
+      statuses: statuses_for_route(route_id, alerts, time)
+    }
+  end
+
+  defp statuses_for_route(route_id, alerts, time) do
+    route_id
+    |> alerts_for_route(alerts)
+    |> alerts_to_statuses(time)
+  end
+
+  # Returns a branch_with_status entry, to be used in the
+  # branches_with_statuses field in groups/2. If no branch_ids are
+  # provided, then uses an empty array.
+  # 
+  defp branch_with_statuses_entry(statuses, branch_ids \\ []) do
+    %{
+      branch_ids: branch_ids,
+      status_entries: statuses
+    }
+  end
+
+  # Given `alerts` and `route_id`, filters out only the alerts
+  # applicable to the given route, using the alert's "informed
+  # entities".
+  defp alerts_for_route(route_id, alerts) do
```

Comment:
suggestion(non-blocking): expanding on my earlier suggestion, this function could potentially be replaced with `Alerts.Match.match(alerts, %InformedEntity{route: route_id})`

Reply:
I'll give this a try again! I'm pretty sure that this is exactly what I tried and couldn't get it to work, but I will give it another go!